packages/netlify-cms-widget-markdown/src/serializers/remarkShortcodes.js

```diff
@@ -10,17 +10,16 @@ export function remarkParseShortcodes({ plugins }) {
 
 function createShortcodeTokenizer({ plugins }) {
   return function tokenizeShortcode(eat, value, silent) {
-    let match;
-    const potentialMatchValue = value.split('\n\n')[0].trimEnd();
-    const plugin = plugins.find(plugin => {
-      match = value.match(plugin.pattern);
-
-      if (!match) {
-        match = potentialMatchValue.match(plugin.pattern);
-      }
-
-      return !!match;
-    });
```

Comment:
@erezrokah I think I understand what was happening here, and what's happening with the image component.

Netlify CMS no longer auto-enables the multi-line flag for all plugin patterns, as of #3089. The image pattern is defined as `/^!\[(.*)\]\((.*?)(\s"(.*)")?\)$/`, which includes `^` and `$` characters. Since the image pattern doesn't have the multi-line flag, `$` is matching the end of the string, not the end of the line. That means any situation in which there is _any_ additional content after the image shortcode (e.g. a file-ending newline, like the example in #3077), the image pattern won't be matched.

I think there are a few paths forward. One would be to add a multiline flag to the image pattern (`/^!\[(.*)\]\((.*?)(\s"(.*)")?\)$/` -> `/^!\[(.*)\]\((.*?)(\s"(.*)")?\)$/m`), or to remove the `^` and `$` tokens, thereby removing the requirement that the entirety of the content needs to exactly match the shortcode. This is most likely insufficient, as there are probably a ton of plugins that have been developed based on this assumption that even if a multi-line flag isn't provided, it's safe to use `$` tokens in plugin patterns.

Another option is to include a more robust version of this single-line fallback in the new implementation:

```javascript
    const trimmedLines = value.split('\n\n').map(line => line.trimEnd());
    const matches = plugins
      .toList()
      .map(plugin => ({
        match: value.match(plugin.pattern) ||
          trimmedLines
            .map(line => line.match(plugin.pattern))
            .filter(match => !!match),
        plugin,
      }))
      .filter(({ match }) => !!match)
      .sort((a, b) => a.match.index - b.match.index);

    const { plugin, match } = matches.get(0) ?? {};
```

With this implementation, we can resolve the ordering issues from the previous implementation, while maintaining the "implicit multi line pattern" behavior.

Thoughts? If this second approach makes sense, I can write up some tests to capture this scenario, ensure that the new implementation is successful, and push up these changes!

Reply:
Ah! There is, in fact, already a test capturing this scenario! It should have been failing on this PR, but the mock `EditorComponent` in `remarkShortcodes.spec.js` was still overriding the flags on `plugin.pattern` to add the multiline flag!

Test: https://github.com/netlify/netlify-cms/blob/master/packages/netlify-cms-widget-markdown/src/serializers/__tests__/remarkShortcodes.spec.js#L28-L32

Stale mock: https://github.com/netlify/netlify-cms/blob/master/packages/netlify-cms-widget-markdown/src/serializers/__tests__/remarkShortcodes.spec.js#L16-L19

Awesome. I'll push my proposed changes, along with the fixed mock!

Reply:
Woo, this makes much more sense now.
Thank you for figuring it out. Do we want to match each line or just the first one as before?

Reply:
I think we have to search all lines; the previous first-line-only implementation was also susceptible to the bug this PR is addressing.

In a first-line-only implementation, if Shortcode1 contains a `$` character, and Shortcode2 does not, we can imagine a `value` where Shortcode1 is found on line 2, and Shortcode2 is found on line 3. When searching for Shortcode1, we won't find it on line 2 (because of the `$`), and we won't find it on the fallback on line one. We will find Shortcode2 on line 3, however. This means we will skip Shortcode1 entirely.